src/bytes_mut.rs

```diff
@@ -819,6 +819,79 @@ impl BytesMut {
         }
     }
 
+    /// Attempts to reclaim the whole allocation of the `BytesMut`.
+    ///
+    /// If the `BytesMut` is empty but the underlying storage has been used before,
+    /// it might be possible to cheaply reclaim capacity as long as there are no
+    /// other `Bytes` or `BytesMut` handles alive.
+    ///
+    /// Returns `true` if the allocation was reclaimed (or reclaiming was not
+    /// necessary, because the underlying storage was empty); returns `false` if the
+    /// `BytesMut` is not empty or there are other outstanding references to the
+    /// underlying storage.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bytes::BytesMut;
+    ///
+    /// let mut buf = BytesMut::with_capacity(64);
+    /// assert_eq!(true, buf.try_reclaim());
+    ///
+    /// buf.extend_from_slice(b"abcd");
+    /// let mut split = buf.split();
+    /// assert_eq!(false, split.try_reclaim());
+    /// assert_eq!(false, buf.try_reclaim());
+    /// drop(buf);
+    /// assert_eq!(false, split.try_reclaim());
+    /// split.clear();
+    /// assert_eq!(true, split.try_reclaim());
+    /// ```
+    #[must_use = "consider BytesMut::reserve if you need a specific capacity"]
+    pub fn try_reclaim(&mut self) -> bool {
+        if !self.is_empty() {
+            return false;
+        }
+
+        let kind = self.kind();
+        if kind == KIND_VEC {
+            unsafe {
+                // Short-circuit if we are already at the start of the vector.
+                let off = self.get_vec_pos();
+                if off == 0 {
+                    return true;
+                }
+                // Otherwise, update info to point at the front of the vector
+                // again and reclaim capacity
+                let base_ptr = self.ptr.as_ptr().sub(off);
+                self.ptr = vptr(base_ptr);
+                self.set_vec_pos(0);
+                self.cap += off;
+                return true;
+            }
+        }
+        let shared: *mut Shared = self.data;
+
+        unsafe {
+            // If there are other references to the underlying storage, we cannot reclaim
+            if !(*shared).is_unique() {
+                return false;
+            }
+            let v = &mut (*shared).vec;
+            let cap = v.capacity();
+            if cap == 0 {
+                // Short-circuit as an empty vector is trivially reclaimed already
+                return true;
+            }
+
+            // Update info to point at start of allocation and reclaim capacity
+            let ptr = v.as_mut_ptr();
+            self.ptr = vptr(ptr);
+            self.cap = cap;
+            true
+        }
+    }
```

Comment:
Instead of such a massive unsafe block, please wrap each unsafe operation in an unsafe block. Same for the `KIND_VEC` case.

Also, please introduce safety comments that explain why the unsafe operations are okay.

Reply:
Done that, thanks!